037-auto-rebalance-cluster-scaling.md

```diff
@@ -0,0 +1,251 @@
+
+## Current situation
+
+Currently, there is no automatic rebalancing when the user scales the cluster up or down.
+It means that:
+
+* after scaling up the cluster, in order to rebalance the cluster and move some replicas to the newly added brokers, the user has to create a `KafkaRebalance` custom resource using the `spec.mode: add-brokers` mode and add the new broker IDs to the the `spec.brokers` field.
+* before scaling down the cluster, in order to rebalance the cluster and move replicas off of the brokers to be removed, the user has to create a `KafkaRebalance` custom resource using the `spec.mode: remove-brokers` mode and add the broker IDs of the brokers to be removed to the `spec.brokers` field.
+
+For more details, the proposal [035](https://github.com/strimzi/proposals/blob/main/035-rebalance-types-scaling-brokers.md) introduced the rebalancing modes mentioned above.
+
+## Motivation
+
+The rebalancing operation during scaling up and down is a manual two step process.
+
+* scaling up the cluster, then running a rebalance.
+* running a rebalance, then scaling down the cluster.
+
+In order to simplify the process, it would be great for the user to deal with just scaling the cluster by changing the `spec.kafka.replicas` field and then having the operator taking care of running a rebalancing operation automatically if requested.
+
+Taking into account the `StrimziPodSets` usage (instead of the `StatefulSets`), which will be the default in the future, a more general use case will be about adding new brokers and/or removing existing ones, not talking just about scaling up and down anymore.
+
+This proposal takes into account this future plan by referring to more general operations like add and remove brokers instead of scaling up and down.
+
+## Proposal
+
+A new `KafkaRebalanceTemplate` custom resource could be a solution allowing the user to define the configuration for the auto-rebalancing but leaving to the operator the creation of the corresponding `KafkaRebalance` custom resource.
+This new CRD should be quite similar to the `KafkaRebalance` one.
+The cluster operator has to copy over the configuration from the template to the actual rebalancing resource.
+Compared to the rejected alternative, of adding the configuration to the `Kafka` custom resource, this will avoid polluting the `Kafka` custom resource and increasing its size.
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaRebalanceTemplate
+metadata:
+  name: my-rebalance-template
+  labels:
+    strimzi.io/cluster: my-cluster
+spec:
+  # NOTE: mode and brokers fields are not available as in the KafkaRebalance,
+  #       they are automatically set on the corresponding KafkaRebalance by the operator
+  goals:
+    - CpuCapacityGoal
+    - NetworkInboundCapacityGoal
+    - DiskCapacityGoal
+    - RackAwareGoal
+    - MinTopicLeadersPerBrokerGoal
+    - NetworkOutboundCapacityGoal
+    - ReplicaCapacityGoal
+  skipHardGoalCheck: true
```

Comment:
`false` is the default in Cruise Control. In this case the rebalance will fail. So, it means that for the scale up use case we'll have a scaled up but not rebalanced cluster, as described in the "error handling" section, the `KafkaRebalance` will be in a `NotReady` state and the user has to deal with it. For the scale down use case, the failed rebalance will stop the scaling down. As above, the user has to deal with a `KafkaRebalance` in `NotReady` state.

Reply:
That is the part I don't fully get. You want to have a separate resource for configuring the rebalance for scale-up / scale-down. But then you anyway mix it with the settings from the Kafka CR. My naive expectation would be that if I have to create this template resource, it will be the single source of truth here and not mix with anything else. Which I guess would mean that if this conflicts with the hard goals, they will be always skipped anyway and this field is not needed.

I wonder if this is intuitive for the user. I would either say that it just uses always the default goals from the Kafka CR. Or that the template is the single source of truth here and the pre-configured goals are always ignored.

Reply:
This is actually how the current `KafkaRebalance` custom resoure and in general Cruise Control works.
From Strimzi documentation about `skipHardGoalCheck` this is what you get:

> Whether to allow the hard goals specified in the Kafka CR to be skipped in optimization proposal generation. This can be useful when some of those hard goals are preventing a balance solution being found. Default is false.

So the user is able to provide goals in his own proposal. There could be some of them as part of the `hard.goals` list but not all of them. It seems that anyway CC takes into account ALL the `hard.goals` (provided by user through the list or the ones in the CC configuration). So it's already mixing things. @kyguy @tomncooper anything I am missing here?

Reply:
I admit it would be pretty silly to provide custom goals that don't include all the hard goals and then set `skipHardGoalsCheck` to `false`. But this matches what we do with the `KafkaRebalance` CR currently. If anything having the default set to `false` is a useful safety net. 

`skipHardGoalsCheck` is one of the many per-rebalance configs that can be set. I think the template should have all the supported configs that are available in the standard `KafkaRebalance` CR. Not least because in future we may want to use the template for some kind of self-healing rebalance proposals so we should try to make it as flexible as possible.

Reply:
+1 @tomncooper